### PR TITLE
Map 197 raid integer array

### PR DESCRIPTION
**Pull request by ChrisC027**
# PR Details
## Description

Updated RGA and MAP encoder, jni wrapper, message builder. 
## Related Jira Key
https://usdot-carma.atlassian.net/browse/MAP-197?atlOrigin=eyJpIjoiMWEyMjgxZTM4ZmM2NGNkZGI2OTBkZWM4YTA3NDAzZDgiLCJwIjoiaiJ9

## Motivation and Context

During integration of RGA base layer it was found out that Road Authority ID type is an [Object Identifier](https://learn.microsoft.com/en-us/windows/win32/seccertenroll/about-object-identifier?redirectedfrom=MSDN), which is an array of integers, smaller or equal to 127.

ISD tool needs to be updated to consider RAID in the format of a list of integers.

## How Has This Been Tested?
Unit tests within the map/rga encoders and message builders have been updated accordingly, and have been verified as passing.

## Types of changes

- [ x ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ x ] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ x ] I have added tests to cover my changes.
- [ x ] All new and existing tests passed.

### CODE REVIEW

File: fedgov-cv-lib-asn1c/src/main/java/rga_wrapper.c [hunk 1]
```diff
@@ -139,67 +138,60 @@ JNIEXPORT jbyteArray JNICALL Java_gov_usdot_cv_rgaencoder_Encoder_encodeRGA(JNIE
...
+		roadAuthorityID->choice.relRdAuthID = relRAIDObjID;
 		rdGeoRefID.rdAuthorityID = roadAuthorityID;
 	} else {
 		roadAuthorityID->present = RoadAuthorityID_PR_NOTHING;
 		rdGeoRefID.rdAuthorityID = NULL;
 	}
 
 	//TODO: SET RAID correctly once implemention is updated in the UI and backend for MAP & RGA.
+	rdGeoRefID.rdAuthorityID = roadAuthorityID;
+	//rdGeoRefID.rdAuthorityID = NULL;
```
Comment (paulbourelly999) [inline fedgov-cv-lib-asn1c/src/main/java/rga_wrapper.c:186]: Please remove any commented out code.

File: fedgov-cv-lib-asn1c/src/main/java/wrapper.c [hunk 2]
```diff
@@ -162,57 +162,56 @@ JNIEXPORT jbyteArray JNICALL Java_gov_usdot_cv_mapencoder_Encoder_encodeMap(JNIE
...
 			jmethodID isRelRdAuthIDExists = (*env)->GetMethodID(env, intersectionClass, "isRelRdAuthIDExists", "()Z");
 			jboolean relRdAuthIDExists = (*env)->CallBooleanMethod(env, intersectionObj, isRelRdAuthIDExists);
 			
 			if (fullRdAuthIDExists)
 			{
 				// Get full RAID
-				jmethodID getFullRdAuthID = (*env)->GetMethodID(env, intersectionClass, "getFullRdAuthID", "()Ljava/lang/String;");
-				jstring fullRdAuthID = (*env)->CallObjectMethod(env, intersectionObj, getFullRdAuthID);
-				const char *fullRdAuthIDStr = (*env)->GetStringUTFChars(env, fullRdAuthID, 0);
+				jmethodID getFullRdAuthID = (*env)->GetMethodID(env, intersectionClass, "getFullRdAuthID", "()[I");
```
Comment (paulbourelly999) [inline fedgov-cv-lib-asn1c/src/main/java/wrapper.c:172]: What is the purpose of the string "()[I"

Comment (ChrisC027) [reply fedgov-cv-lib-asn1c/src/main/java/wrapper.c:172]: GetMethodID JNI Function requires a type signature: https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetMethodID

"()[I" denotes a array of integers: https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/types.html

File: fedgov-cv-lib-asn1c/src/main/java/rga_wrapper.c [hunk 3]
```diff
@@ -139,67 +138,60 @@ JNIEXPORT jbyteArray JNICALL Java_gov_usdot_cv_rgaencoder_Encoder_encodeRGA(JNIE
...
+		const int *fullRAID = (*env)->GetIntArrayElements(env, fullRdAuthID, NULL);
 
-		jsize len = (*env)->GetArrayLength(env, fullRdAuthID);
-		size_t fullRdAuthIDStrLen = len;
+		size_t fullRAIDLen = (*env)->GetArrayLength(env, fullRdAuthID);
 
-		ASN__PRIMITIVE_TYPE_t primType;
-		primType.buf = (uint8_t *)calloc(1, (fullRdAuthIDStrLen + 1));
+		OBJECT_IDENTIFIER_t fullRAIDObjID;
+		fullRAIDObjID.buf = (uint8_t *)calloc(1, (fullRAIDLen + 1));
```
[the remark this answers is not in the document]
Comment (DokurOmkar) [reply fedgov-cv-lib-asn1c/src/main/java/rga_wrapper.c:146]: From the documentation [here](https://learn.microsoft.com/en-us/windows/win32/seccertenroll/about-object-identifier?redirectedfrom=MSDN), OBJECT IDENTIFIER is encoded by following these rules:

1. The first two nodes of the OID are encoded onto a single byte. The first node is multiplied by the decimal 40 and the result is added to the value of the second node.
2. Node values less than or equal to 127 are encoded on one byte.
3. Node values greater than or equal to 128 are encoded on multiple bytes. Bit 7 of the leftmost byte is set to one. Bits 0 through 6 of each byte contains the encoded value.

The code needs to be updated as follows:

- Start by initializing a encodedRAID pointer (uint8_t) that is of length greater than integer array (say twice the length of the integer array)
- After checking length of the array is greater than 2, implement logic  that follows the rule 1. Now, assign this to the first position of the encodedRaid
- For the next integers, loop through in a for loop and check if each of the integer is greater than 127 or not.
- If it is less than 127, assign it to the next encodedRAID position without changing anything (Step 2)
- If it is greater than 127, follow step 3 and assign the result into the next encodedRAID position.
- Now, assign encodedRAID to the buffer with size as the length of the encodedRAID .

File: fedgov-cv-lib-asn1c/src/main/java/wrapper.c [hunk 4]
```diff
@@ -162,57 +162,56 @@ JNIEXPORT jbyteArray JNICALL Java_gov_usdot_cv_mapencoder_Encoder_encodeMap(JNIE
...
+				jintArray fullRdAuthID = (*env)->CallObjectMethod(env, intersectionObj, getFullRdAuthID);
+				const int *fullRAID = (*env)->GetIntArrayElements(env, fullRdAuthID, NULL);
 
-				size_t fullRdAuthIDStrLen = strlen(fullRdAuthIDStr);
+				size_t fullRAIDLen = (*env)->GetArrayLength(env, fullRdAuthID);
 
-				ASN__PRIMITIVE_TYPE_t primType;
-				primType.buf = (uint8_t *)calloc(1, (fullRdAuthIDStrLen + 1));
+				OBJECT_IDENTIFIER_t fullRAIDObjID;
```
Comment (DokurOmkar) [inline fedgov-cv-lib-asn1c/src/main/java/wrapper.c:178]: Follow the same as above for OBJECT IDENTIFIER

**Review by DokurOmkar — CHANGES_REQUESTED** [review]

File: fedgov-cv-message-builder/src/test/java/gov/usdot/cv/msg/builder/IntersectionSituationDataBuilderTest.java [hunk 5]
```diff
@@ -135,21 +135,21 @@ public void mapMessageWithNodeSpeedLimitTest() {
     @Test
     public void rgaBaseLayerMessageWithFramePlusRGATypeTest() {
         String filePath = "src/test/resources/samplerga.json";
-        String expectedHexString = "00000a66d9b231ea6b5f708e13f410401808101807f7e820b420020690";
+        String expectedHexString = "00000a66d9b231ea6b5f708e13f4105008a80c0602040601fdfa082d080081a400";
         runIntersectionSituationDataBuilderTests(filePath, expectedHexString);
     }
```
Comment (DokurOmkar) [inline fedgov-cv-message-builder/src/test/java/gov/usdot/cv/msg/builder/IntersectionSituationDataBuilderTest.java:141]: Add more unit test cases to test the Relative Road Authority ID and Mapped GeomID. Make sure to test a node greater than 127 for Relative Road Authority ID.

**Review by DokurOmkar — CHANGES_REQUESTED** [review]
Please also make sure to check SonarCloud and fix any issues.

**Review by DokurOmkar — APPROVED** [review]

**Pull request merged**